### PR TITLE
[eas-cli] Add channel edit list and view commands

### PR DESCRIPTION
**Pull request by byCedric**
# Why

This adds the 3 commands from [the CLI design doc](https://www.notion.so/expo/CLI-Design-bc3d71bad49c4b8eab7f7b00e6b867bf#333b0753046e4279906ed9ff413cd906):
- [`channel:list`](https://linear.app/expo/issue/ENG-254/implement-channellist)
- [`channel:view`](https://linear.app/expo/issue/ENG-256/implement-channelview)
- [`channel:edit`](https://linear.app/expo/issue/ENG-255/implement-channeledit)

# How

- Moved existing GraphQL mutations from `channel:create` to `graphql/mutations/ChannelMutation.ts`
- Added new `graphq/queries/ChannelQuery.ts`, with list and get by id queries
- Created the 3 commands `channel:view`, `channel:list`, and `channel:edit`

# Test Plan

If you run these commands from a project, you should be able to test all of the 3 commands:

- `$ eas channel:list`, to see no results
- `$ eas channel:create <name>`, to add at least 1 channel + branch

- `$ eas branch:create <otherName>`, to add another branch
- `$ eas branch:publish`, to add an update to this branch

- `$ eas channel:edit`, to update the channel to the new branch
- `$ eas channel:view` to see a single channel

### CODE REVIEW

File: packages/eas-cli/src/commands/channel/create.ts [hunk 1]
```diff
@@ -127,10 +83,14 @@ export default class ChannelCreate extends Command {
...
     }
 
-    const newChannel = await createUpdateChannelOnAppAsync({
+    const newChannel = await ChannelMutation.createForAppAsync({
       appId: projectId,
-      channelName,
-      branchId,
+      name: channelName,
+      // Point the new channel at a branch with its same name.
+      branchMapping: JSON.stringify({
```
Comment (byCedric) [inline packages/eas-cli/src/commands/channel/create.ts:90]: We should probably make a utility for this, to translate a set of (+ possible rollout options?) to a `branchMapping` JSON.

File: packages/eas-cli/src/commands/channel/edit.ts [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    });
+
+    const { id: channelId } = await ChannelQuery.byNameForAppAsync({
+      appId: projectId,
+      name: channelName,
+    });
+
+    const channel = await ChannelMutation.updateBranchMappingAsync({
+      channelId,
+      branchMapping: JSON.stringify({
```
Comment (byCedric) [inline packages/eas-cli/src/commands/channel/edit.ts:95]: Same here:
> We should probably make a utility for this, to translate a set of (+ possible rollout options?) to a `branchMapping` JSON.

File: packages/eas-cli/src/commands/channel/list.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+      const branches = channel.updateBranches;
+      const update = branches
+        .map(branch => branch.updates)
+        .flat()
+        .sort((a, z) => z.createdAt - a.createdAt)[0];
+
+      table.push([
+        channel.name,
+        // todo: replace with branch mapping
+        branches.map(branch => branch.name).join(', '),
```
Comment (byCedric) [inline packages/eas-cli/src/commands/channel/list.ts:62]: I'm not sure how I can translate that JSON to something like `v1.1 (10%) / v1.0 (90%)`. So this should probably be changed.

File: packages/eas-cli/src/commands/channel/list.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    const table = new Table({
+      head: ['channel', 'branch', 'update', 'created-at', 'actor'],
+      wordWrap: true,
+    });
+
+    for (const channel of channels) {
+      const branches = channel.updateBranches;
+      const update = branches
```
Comment (byCedric) [inline packages/eas-cli/src/commands/channel/list.ts:54]: This is also a hard one for me, related to the branchmapping.

- If there is only 1 branch, we can grab the latest update from that branch
- If there are multiple branches (with a rollout setup), _what update do we need to display here?_

File: packages/eas-cli/src/graphql/queries/ChannelQuery.ts [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    const data = await withErrorHandlingAsync(
+      graphqlClient
+        .query<ListChannelsForAppQuery>(
+          gql`
+            query ListChannelsForApp($appId: String!, $offset: Int!, $limit: Int!) {
+              app {
+                byId(appId: $appId) {
+                  id
+                  name
+                  updateChannels(offset: $offset, limit: $limit) {
```
Comment (byCedric) [inline packages/eas-cli/src/graphql/queries/ChannelQuery.ts:76]: The offset/limiting here is a bit weird. There are 3 within this query, so I'm not 100% sure if this is correct.

File: packages/eas-cli/src/commands/channel/edit.ts [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+import {
+  findProjectRootAsync,
+  getBranchByNameAsync,
+  getProjectAccountNameAsync,
+} from '../../project/projectUtils';
+import { promptAsync } from '../../prompts';
+
+export default class ChannelEdit extends Command {
+  static hidden = true;
+  static description = 'Edit a channel on the current project.';
```
Comment (jkhales) [inline packages/eas-cli/src/commands/channel/edit.ts:18]:
```suggestion
  static description = 'Point a channel at a new branch.';
```

File: packages/eas-cli/src/graphql/queries/ChannelQuery.ts [hunk 7]
```diff
@@ -0,0 +1,116 @@
+import gql from 'graphql-tag';
```
Comment (jkhales) [inline packages/eas-cli/src/graphql/queries/ChannelQuery.ts:1]: We're trying to move away from the centralized `graphql` dir and put the graphQL in the same file if possible. Would it be possiblle to move these function inside as well?

**Review by jkhales — COMMENTED** [review]
Looking good, would we be able to move the graphQL functions into the command files?

Left a bunch of small comments as well.

File: packages/eas-cli/src/graphql/queries/ChannelQuery.ts (hunk 7 above)
Comment (byCedric) [reply packages/eas-cli/src/graphql/queries/ChannelQuery.ts:1]: ah, gotcha, I thought this was the way to go. I'll move them back!

File: packages/eas-cli/src/commands/channel/edit.ts [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+    if (!branchName) {
+      const validationMessage = 'branch name may not be empty.';
+      if (jsonFlag) {
+        throw new Error(validationMessage);
+      }
+      ({ name: branchName } = await promptAsync({
+        type: 'text',
+        name: 'name',
+        message: 'What branch should it change to?',
```
[the remark this answers is not in the document]
Comment (byCedric) [reply packages/eas-cli/src/commands/channel/edit.ts:147]: Which one do you prefer?

**Currently:**
Screenshot 2021-02-26 at 01 14 27

**My personal preference:**
Screenshot 2021-02-26 at 01 12 44

File: packages/eas-cli/src/commands/channel/view.ts [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+  static flags = {
+    json: flags.boolean({
+      description: 'print output as a JSON object with the channel ID, name and branch mapping.',
+      default: false,
+    }),
+  };
+
+  async run() {
+    let {
+      args: { name: channelName },
```
[the remark this answers is not in the document]
Comment (byCedric) [reply packages/eas-cli/src/commands/channel/view.ts:82]: I named it `channelName` to be consistent with `appId` and `branchName`. But we can switch it if you want

[the remark this answers is not in the document]
Comment (jkhales) [reply packages/eas-cli/src/commands/channel/view.ts:82]: ah, makes sense.

File: packages/eas-cli/src/commands/channel/edit.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (jkhales) [reply packages/eas-cli/src/commands/channel/edit.ts]: Those both look great. Either work!

File: packages/eas-cli/src/commands/channel/list.ts [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+      return;
+    }
+
+    const table = new Table({
+      head: ['channel', 'branch', 'update', 'message', 'created-at', 'actor'],
+      wordWrap: true,
+    });
+
+    for (const channel of channels) {
+      // todo: refactor when multiple branches per channel are available
```
Comment (jkhales) [inline packages/eas-cli/src/commands/channel/list.ts:106]:
```suggestion
      // todo (cedric): refactor when multiple branches per channel are available
```
not sure what the custom is on this repo, but signing todo's like this helps keep them from getting forgotten.

Comment (jkhales) [reply packages/eas-cli/src/commands/channel/list.ts:106]: also capitalized TODO is a bit easier to grep

File: packages/eas-cli/src/commands/channel/view.ts [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+      return;
+    }
+
+    const table = new Table({
+      head: ['branch', 'update', 'message', 'created-at', 'actor'],
+      wordWrap: true,
+    });
+
+    for (const branch of channel.updateBranches) {
+      // todo: refactor when multiple branches per channel are available
```
Comment (jkhales) [inline packages/eas-cli/src/commands/channel/view.ts:125]:
```suggestion
      // TODO (cedric): refactor when multiple branches per channel are available
```

File: packages/eas-cli/src/commands/channel/list.ts [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+                        }
+                      }
+                    }
+                  }
+                }
+              }
+            }
+          }
+        `,
+        { appId, limit }
```
Comment (jkhales) [inline packages/eas-cli/src/commands/channel/list.ts:57]:
```suggestion
        { appId, limit: CHANNEL_LIMIT }
```

File: packages/eas-cli/src/commands/channel/list.ts [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+import { findProjectRootAsync, getProjectAccountNameAsync } from '../../project/projectUtils';
+
+const CHANNEL_LIMIT = 10_000;
+
+async function getAllUpdateChannelForAppAsync({
+  appId,
+  limit = CHANNEL_LIMIT,
+}: {
+  appId: string;
+  limit?: number;
```
Comment (jkhales) [inline packages/eas-cli/src/commands/channel/list.ts:20]:
```suggestion
}: {
  appId: string;
```

**Review by jkhales — APPROVED** [review]
🔥

File: packages/eas-cli/src/commands/channel/view.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (byCedric) [reply packages/eas-cli/src/commands/channel/view.ts:82]: I think it also helps you "understand" the values, regardless of looking at the command name. But again, I'm open to this 😄

**Pull request merged**